Link to Tags using sluggedName instead of the name itself.

MatejBalantic opened the pull request:
* this allows tag names to contain any character, including spaces etc, thus validation of the name is now removed
* slugged name is automatically generated in the same way as for the questions
* the change is backwards-compatible as all the names are transferred to suggedName upon migration

Notice: this passes unit tests, however will not build on Travis until https://github.com/caelum/mamute/pull/134 is merged

Closes #120
Closes #119

leocwolter commented on src/main/java/org/mamute/controllers/ListController.java:
```diff
@@ -159,9 +159,9 @@ public void unanswered(Integer p) {
 	
 	@Get
-	public void withTag(String tagName, Integer p, boolean semRespostas) {
+	public void withTag(String tagUriName, Integer p, boolean semRespostas) {
```
that variable `semRespostas` should be `withoutAnswers`. It's in portuguese ):

MatejBalantic replied:
I noticed, but resisted an urge to fix it :) I'll change it with other things

leocwolter commented on src/main/java/org/mamute/controllers/RankingController.java:
```diff
@@ -34,8 +34,8 @@ public void rank(Integer p) {
 	@Get
-	public void tagRank(String tagName){
-		Tag tag = tags.findByName(tagName);
+	public void tagRank(String tagUriName){
```
Better not to change controllers methods parameter's names. They are used at [`routes.properties`](https://github.com/caelum/mamute/blob/master/src/main/resources/routes.properties) files and someone could overwrite those.

MatejBalantic replied:
Thanks, good point. Will revert the routes back

leocwolter commented on src/test/java/org/mamute/dao/TagDAOTest.java:
```diff
@@ -141,6 +150,25 @@ public void should_not_save_repeated_tags() throws Exception {
+
+    @Test
+    public void should_define_correct_slugged_url()
+    {
```
we open curly braces at the same line as the method signature :)

MatejBalantic replied:
My apologies, I'm coming from Objective-C :) Will fix it of course

leocwolter commented:
Since we support the configurations below at mamute instances, the user can use a dash to split tags ):
```
# Regular expression to split tags inside Java
tags.splitter.regex = \\,
# Char used to split tags inside the JSP forms. It must matchs with the tags.splitter.regex property
tags.splitter.char = ,
# Regex of allowed characters for tag names
tags.sanitizer.regex = [a-zA-Z0-9-]
```
I think we should use them when generating the slugged tags(maybe the Normalizer should be customizable too)

MatejBalantic commented:
I noticed that but I didn't see any case where this would matter. Is there
any? Can you provide an example of string that could cause issues?


On Thursday, January 15, 2015, Leonardo Cesar Wolter <
notifications@github.com> wrote:

> Since we support the configurations below at mamute instances, the user
> can use a dash to split tags ):
>
> # Regular expression to split tags inside Java
> tags.splitter.regex = \\,
> # Char used to split tags inside the JSP forms. It must matchs with the tags.splitter.regex property
> tags.splitter.char = ,
> # Regex of allowed characters for tag names
> tags.sanitizer.regex = [a-zA-Z0-9-]
>
> I think we should use them when generating the slugged tags(maybe the
> Normalizer should be customizable too)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/caelum/mamute/pull/135#issuecomment-70033562>.
>


-- 

*Matej Balantič*

T +49 171 2976495
W http://about.me/Matej.Balantic

leocwolter commented:
I could use dash to split and allow spaces on my tags: java ee-jpa
The final result would be java-ee and jpa. If I decide to update those tags, it would be shown as java-ee-jpa :)

MatejBalantic commented:
That shouldn't be true as only the sluggedName would contain the dash, not
the name. So you would still end up with
java ee
jpa

after insert which would work fine when updating. Maybe I just don't
understand, can you write a unit test which would cause the issue you are
talking about?

On Thursday, January 15, 2015, Leonardo Cesar Wolter <
notifications@github.com> wrote:

> I could use dash to split and allow spaces on my tags: java ee-jpa
> The final result would be java-ee and jpa. If I decide to update those
> tags, it would be shown as java-ee-jpa :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/caelum/mamute/pull/135#issuecomment-70085693>.
>


-- 

*Matej Balantič*

T +49 171 2976495
W http://about.me/Matej.Balantic

leocwolter commented on src/main/java/org/mamute/model/Tag.java:
```diff
@@ -28,7 +30,11 @@
+
+    @Type(type = "text")
+    @NotEmpty
+    private String sluggedName;
```
that one should have unique and not null constraint  in it too, just in case :)

leocwolter commented:
Oh, we are keeping the name and the sluggedName, got it. Sorry, I thought you wanted to normalize the name of the tags :)

leocwolter commented on src/main/java/org/mamute/model/Tag.java:
```diff
@@ -48,7 +54,8 @@ public Tag() {
 		this.name = name.toLowerCase();
-		this.description = description;
+        this.sluggedName = toSlug(name);
+        this.description = description;
```
We use tab for indentation(with size of 4 blank spaces), can you configure your editor for that?

MatejBalantic replied:
Of course. Should this project maybe have a wiki with coding standards and editor settings explained? I imagine most of them are industry standard in Java world, but not everybody are from Java world (khm, people like me:)

csokol replied:
I think we can create a "Contributing" section in our readme or a question at meta.mamute.org for that...

leocwolter replied:
Yeah, or maybe we can add the eclipse / intellij code style file at git

leocwolter commented on src/main/java/org/mamute/model/Tag.java:
```diff
@@ -57,12 +64,7 @@ public String getName() {
-			e.printStackTrace();
-		}
-		return null;
+		return this.sluggedName;
```
Could you test tags like `c++` and `c#` please? The slugged name cannot be just `c` since it would conflict

leocwolter commented on src/main/resources/mamute.properties:
```diff
@@ -14,8 +14,6 @@ vraptor.errorcontrol.date.joda.pattern = dd/MM/yyyy
 # Char used to split tags inside the JSP forms. It must matchs with the tags.splitter.regex property
 tags.splitter.char = ,
-# Regex of allowed characters for tag names
-tags.sanitizer.regex = [a-zA-Z0-9-]
```
Sorry, I didn't get it, why did you remove this?

MatejBalantic replied:
I removed it as we are no longer validating the tag names (not needed due
to slugs). Will have to rethink the approach to get c++ and c# working
together though. Thanks for pointing this out.

On Thursday, January 15, 2015, Leonardo Cesar Wolter <
notifications@github.com> wrote:

> In src/main/resources/mamute.properties
> <https://github.com/caelum/mamute/pull/135#discussion-diff-23029896>:
>
> > @@ -14,8 +14,6 @@ vraptor.errorcontrol.date.joda.pattern = dd/MM/yyyy
> >  tags.splitter.regex = \\,
> >  # Char used to split tags inside the JSP forms. It must matchs with the tags.splitter.regex property
> >  tags.splitter.char = ,
> > -# Regex of allowed characters for tag names
> > -tags.sanitizer.regex = [a-zA-Z0-9-]
>
> Sorry, I didn't get it, why did you remove this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/caelum/mamute/pull/135/files#r23029896>.
>


-- 

*Matej Balantič*

T +49 171 2976495
W http://about.me/Matej.Balantic

leocwolter replied:
We could just use the Normalizer to url-encode those things, I guess it makes sense on every case we normalize a string, but maybe we should rename Normalizer to make this more explicit

MatejBalantic replied:
We need to transliterate characters like čžš to czs, and only than we could url-encode non-transliteratable characters like # and +. I'll prepare something and commit it here.

leocwolter commented on src/test/java/org/mamute/model/TagTest.java:
```diff
@@ -0,0 +1,42 @@
+    {
+        assertEquals(subject.getUriName(), "tag-with-specialchars");
+    }
+}
```
same here {}  :)

MatejBalantic replied:
I'll check them all out, sorry about that again.

MatejBalantic commented:
Hi, so I amended the pull requests:
* addressed the review comments regarding indentation and brackets, where I noticed it at least,
* slugging for tags is now done by normalizing alphabetic characters and then url encoding special characters. Also added tests for C# and c++

leocwolter commented on src/main/java/org/mamute/controllers/TagPageController.java:
```diff
@@ -47,24 +47,24 @@ public void editTagPage(String tagName, MarkedText about){
 		}
 		tagPages.save(tagPage);
-		result.redirectTo(this).showTagPage(tagPage.getTagName()); 
+		result.redirectTo(this).showTagPage(tagPage.getTag().getUriName());
```
Since we will always need `tagPage.getTag().getUriName()` to get the uri, we could encapsulate this inside tagPage. Something like `tagPage.getTagUriName()` or a better name :P

MatejBalantic replied:
I can change this if you will insist. I personally prefer keeping the structure of the relations on the model as-is, don't like to flatten it, as it's harder to understand the relationships there. I mean, that why the models are there, right :) But if you feel strongly about it, I can change it

leocwolter replied:
Well, the problem is if `getTag` starts to receive an argument at some point, we would need to pass this new argumento at every place we call `tagPage.getTag("argument").getUriName()`.
But I understand your point, in that case we could just use a variable to keep this value so we don't need to call it multiple times :)

MatejBalantic replied:
Since this are DOMs getTag will probably never receive any arguments - it's a to-one relationship, so what could the argument be? It also feels as over-designing, but I'll do it since I don't feel so strongly about it :)

leocwolter commented on src/main/java/org/mamute/dao/TagDAO.java:
```diff
@@ -88,11 +91,13 @@ public boolean hasAbout(Tag tag) {
 
 	public Tag saveIfDoesntExists(Tag newTag) {
-		Tag existingTag = findByName(newTag.getName());
+		Tag existingTag = findBySluggedName(newTag.getUriName());
```
Better to use the same name for both: "sluggedName" or "uriName"

MatejBalantic replied:
will do

leocwolter commented on src/main/java/org/mamute/infra/NormalizerBrutal.java:
```diff
@@ -1,18 +1,34 @@
 		if(input == null || input.isEmpty()) return "";
-		Pattern NONLATIN = Pattern.compile("[^\\w-]");
+
+		String nonLatinPattern = urlEncodeNonNormalizableCharacter ? "[^\\P{M}]" : "[^\\w-]";
```
Sorry, but why was it necessary? Didn't get it

MatejBalantic replied:
You are right that I didn't do the best job explaining what is happening.

So this is what we want:
- normalize characters like čćšž etc. (also called base-glyph, they match with P{M} in unicode)
- NOT normalize characters like  + or # or ! or . and rather URL encode them

If we would normalize character like #, it would first be decomposed to several characters and than URL encoded, which would result in something that is in fact not URL-encoded representation of a hash. 

In case we don't want to URL encode, then we SHOULD normalize all characters and strip out non ASCII representation of them at the end (thus \w)

leocwolter replied:
Oh, okay, it makes sense, sorry :)

leocwolter commented on src/main/java/org/mamute/infra/NormalizerBrutal.java:
```diff
@@ -1,18 +1,34 @@
+		if (urlEncodeNonNormalizableCharacter) {
+			try {
+				slug = URLEncoder.encode(slug, "UTF-8");
+			} catch (UnsupportedEncodingException e) {	}
```
`throw new IllegalArgumentException(e)` just in case :)

MatejBalantic replied:
will do

leocwolter commented on src/main/java/org/mamute/model/Tag.java:
```diff
@@ -24,45 +26,46 @@
+    @Column(unique = true, nullable = false)
+    @Type(type = "text")
+    @NotEmpty
+    private String sluggedName;
```
identation :(

leocwolter commented on src/main/java/org/mamute/model/Tag.java:
```diff
@@ -24,45 +26,46 @@
 		this.name = name.toLowerCase();
-		this.description = description;
+        this.sluggedName = toSlug(name, true);
+        this.description = description;
```
identation :(

leocwolter commented on src/test/java/org/mamute/dao/TagDAOTest.java:
```diff
@@ -141,12 +172,28 @@ public void should_not_save_repeated_tags() throws Exception {
+        assertNotNull(tags.findBySluggedName("java"));
+        assertNotNull(tags.findBySluggedName("ruby"));
+        assertNotNull(tags.findBySluggedName("c%2b%2b"));
+    }
```
better to assert if the returned tag is actually `c++`

MatejBalantic replied:
Good catch

leocwolter commented on src/test/java/org/mamute/dao/TagDAOTest.java:
```diff
@@ -88,17 +96,32 @@ public void should_get_main_tags_of_the_provided_user() throws Exception {
+	public void should_get_existing_tags_with_different_name_different_slug() throws Exception {
+		List<Tag> found = tags.findAllDistinct(asList("C#", "C++"));
+		assertEquals(found.size(), 2);
+	}
```
the test `should_not_repeat_tags_with_similar_name` already covers that, no?

MatejBalantic replied:
Good catch!

leocwolter commented on src/test/java/org/mamute/dao/TagDAOTest.java:
```diff
@@ -88,17 +96,32 @@ public void should_get_main_tags_of_the_provided_user() throws Exception {
+		assertEquals(cplus.getName(), tagsNames.get(1));
+		assertEquals(weirdCharacters.getName(), tagsNames.get(2));
+        assertEquals(java.getName(), tagsNames.get(3));
+        assertEquals(ruby.getName(), tagsNames.get(4));
```
you could use hamcrest matchers to make it easier to read here. Something like:
```
assertThat(tagNames, Matchers.containsInAnyOrder(java.getName(), ruby.getName()....)
```

MatejBalantic replied:
Thanks, done

MatejBalantic commented:
I've addressed the review comments, please review when you can

csokol commented on src/test/java/org/mamute/infra/NormalizerBrutalTest.java:
```diff
@@ -27,6 +32,9 @@ public void should_remove_question_marks_and_bangs() {
 		assertEquals("como-faz-para-normalizar-uma-string", slug);
+
+		String slugEncoded = NormalizerBrutal.toSlug(text, true);
+		assertEquals("como-faz-para-normalizar-uma-string%3f%21", slugEncoded);
```
If we start url encoding like this, we will lose the friendly urls we have for questions (example: http://demo.mamute.org/297-what-happens-if-i-use-strange-chars-on-the-title). 

Maybe we should keep the old normaliser and create a normaliser only for tags. What do you think, @MatejBalantic?

JamesSullivan replied:
Apologies for jumping in. The current pull request is not usable with Asian languages. I would think just using URL encoding is the simplest and covers the most cases for the least work particularly as browsers now [handle the display of Unicode urls](http://stackoverflow.com/questions/2742852/unicode-characters-in-urls) better. There are a few characters such as "?" that are still displayed as % hex code in the url input box but there are good security reasons for that. Realizing this is a question site and most titles will end in a question mark that would show up as %3F  perhaps the only time a character should be stripped is when the title ends in a question mark.

JamesSullivan replied:
We can support non-latin languages by changing one line of code and adding another in NormalizerBrutal. 
[NormalizerBrutal.java.txt](https://github.com/caelum/mamute/files/95251/NormalizerBrutal.java.txt)

All tests pass except for this one

felipeweb commented:
closing here see https://github.com/caelum/mamute/issues/119#issuecomment-192878835 for more details

JamesSullivan commented:
This fix (disallowing most characters) although simpler does not allow Japanese or Chinese tags, etc. The updated brutalizer and tag slugging would have allowed any Unicode language.
